src/AuthenticationRoute.ts

```diff
@@ -0,0 +1,305 @@
+  ? SingleStrategyCallback
+  : MultiStrategyCallback;
+
+const Unhandled = Symbol.for("passport-unhandled");
+
+export class AuthenticationRoute<StrategyNames extends string | string[]> {
+  readonly options: AuthenticateOptions;
+  readonly strategies: string[];
+  readonly isMultiStrategy: boolean;
+
+  constructor(
+    readonly authenticator: Authenticator, // aggregator instance that owns the chain of strategies
+    strategyOrStrategies: StrategyNames, // list of strategies this handler tries
+    options?: AuthenticateOptions, // options governing behaviour of strategies
+    readonly callback?: AuthenticateCallback<StrategyNames> // optional custom callback to process the result of the strategy invocations
+  ) {
+    this.options = options || {};
+
+    // Cast `name` to an array, allowing authentication to pass through a chain of strategies.  The first strategy to succeed, redirect, or error will halt the chain.  Authentication failures will proceed through each strategy in series, ultimately failing if all strategies fail.
+    // This is typically used on API endpoints to allow clients to authenticate using their preferred choice of Basic, Digest, token-based schemes, etc. It is not feasible to construct a chain of multiple strategies that involve redirection (for example both Facebook and Twitter), since the first one to redirect will halt the chain.
+    if (Array.isArray(strategyOrStrategies)) {
+      this.strategies = strategyOrStrategies;
+      this.isMultiStrategy = false;
+    } else {
+      this.strategies = [strategyOrStrategies as string];
+      this.isMultiStrategy = false;
+    }
+  }
+
+  handler = async (request: FastifyRequest, reply: FastifyReply) => {
+    if (!request.passport) {
+      throw new Error("passport.initialize() plugin not in use");
+    }
+    // accumulator for failures from each strategy in the chain
+    const failures: FailureObject[] = [];
+
+    for (const strategy of this.strategies) {
+      try {
+        return await this.attemptStrategy(failures, strategy, request, reply);
+      } catch (e) {
+        if (e == Unhandled) {
+          continue;
+        } else {
+          throw e;
+        }
+      }
+    }
+
+    return this.onAllFailed(failures, request, reply);
+  };
+
+  async attemptStrategy(failures: FailureObject[], name: string, request: FastifyRequest, reply: FastifyReply) {
+    const prototype = this.authenticator.strategy(name);
+    if (!prototype) {
+      throw new Error(`Unknown authentication strategy ${name}!`);
+    }
+
+    const strategy = Object.create(prototype) as Strategy;
+
+    // This is a messed up way of adapting passport's API to fastify's async world. We create a promise that the strategy's per-call functions close over and resolve/reject with the result of the strategy. This augmentation business is a key part of how Passport strategies expect to work.
+    return new Promise((resolve, reject) => {
```

Comment:
I'm not really fond of this approach of allocating new functions every time. Can we have a prototype-based approach where the various methods are added to the prototype instead and we just set some state into the strategy? (Maybe we can add a few symbol-based properties)

Reply:
I too don't like it at all, it's super slow unnecessarily. These functions are part of the Passport interface exposed to strategies (see https://github.com/jaredhanson/passport-http-bearer/blob/master/lib/strategy.js#L86 for example) though so the API of them can't be changed at all, and the prototype chain that the strategies use is also not within this library's control as that's governed by the strategy's code itself. 

This same style is how Passport works internally and I thought it was probably best to mirror that so that the strategies are running in an as-close-as-possible environment to a normal `passport` host. I got rid of a bunch of other inline function allocations and convoluted callback-ery from the ported source, but this is what I was left with because it's the Passport API. 

The only alternative I came up with was to assign these functions to a strategy object once when it was registered and then have them call back into the current attempt via a global, but I felt like that would be hard to get right and risks diverging from passport and breaking strategies unnecessarily.

I kind of think that what is really broken here is Passport's API "spec" and that if folks want really good performance they should use the other cleanroom auth implementations that were designed with performance in mind, whereas this library is more for lazy less performance sensitive folks like me who want to take advantage of the passport strategy ecosystem.

Reply:
As far as I remember, a long time ago when I check this out, it was exactly as @airhorns said. This is the only way (or maybe the more reliable) to implement this since it is the passport API.